File: ezkey-tests/src/test/java/org/ezkey/tests/util/DatabaseHelper.java
/*
 * Ezkey - Open Source MFA/Passkey Alternative
 *
 * Copyright (c) 2025 Ezkey contributors
 * Licensed under the MIT License. See LICENSE file in the project root for full license information.
 *
 * Utility: DatabaseHelper
 * Description: Helper for opportunistic database access in E2E tests
 */

package org.ezkey.tests.util;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.List;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Helper class for opportunistic database access in E2E tests.
 *
 * <p>Provides methods to query and update PostgreSQL database directly via docker exec, avoiding
 * API authentication requirements when appropriate. This is suitable for functional E2E tests in
 * clean room Docker environment.
 *
 * <p><b>Usage Context:</b> Use for:
 *
 * <ul>
 *   <li>State verification (faster than API calls)
 *   <li>Finding existing entities before creating new ones
 *   <li>Cleanup operations
 *   <li>Resetting state for test idempotence
 * </ul>
 *
 * <p><b>Database Configuration:</b> Automatically detects PostgreSQL container:
 *
 * <ul>
 *   <li>HA mode: "ezkey-postgres-ha" (when HA containers are present)
 *   <li>Standard mode: "ezkey-postgres" (default)
 * </ul>
 *
 * <p>Database: "ezkey_db", User: "postgres"
 *
 * @since 2025
 */
public class DatabaseHelper {

  private static final Logger log = LoggerFactory.getLogger(DatabaseHelper.class);

  private static final String DOCKER_CONTAINER_STANDARD = "ezkey-postgres";
  private static final String DOCKER_CONTAINER_HA = "ezkey-postgres-ha";
  private static final String DATABASE = "ezkey_db";
  private static final String USER = "postgres";

  /**
   * Detects which PostgreSQL container to use based on environment (standard or HA mode).
   *
   * @return Container name to use for database operations
   */
  private String detectPostgresContainer() {
    // Check if HA mode container exists
    if (containerExists(DOCKER_CONTAINER_HA)) {
      log.debug("HA mode detected: Using PostgreSQL container {}", DOCKER_CONTAINER_HA);
      return DOCKER_CONTAINER_HA;
    }

    // Fallback to standard mode
    if (containerExists(DOCKER_CONTAINER_STANDARD)) {
      log.debug("Standard mode detected: Using PostgreSQL container {}", DOCKER_CONTAINER_STANDARD);
      return DOCKER_CONTAINER_STANDARD;
    }

    // Default to standard (will fail with clear error if container doesn't exist)
    log.warn(
        "Neither {} nor {} found, defaulting to {}",
        DOCKER_CONTAINER_HA,
        DOCKER_CONTAINER_STANDARD,
        DOCKER_CONTAINER_STANDARD);
    return DOCKER_CONTAINER_STANDARD;
  }

  /**
   * Checks if a Docker container exists.
   *
   * @param containerName Container name to check
   * @return true if container exists, false otherwise
   */
  private boolean containerExists(String containerName) {
    try {
      ProcessBuilder processBuilder = new ProcessBuilder("docker", "inspect", containerName);
      processBuilder.redirectErrorStream(true);
      Process process = processBuilder.start();

      // Consume output to avoid blocking
      try (BufferedReader reader =
          new BufferedReader(new InputStreamReader(process.getInputStream()))) {
        while (reader.readLine() != null) {
          // Consume output
        }
      }

      int exitCode = process.waitFor();
      return exitCode == 0;
    } catch (Exception e) {
      log.debug("Container {} does not exist: {}", containerName, e.getMessage());
      return false;
    }
  }

  /**
   * Executes a SQL query and returns the result as a list of strings (one per row).
   *
   * @param sqlQuery SQL query to execute
   * @return List of result rows (trimmed strings), empty list if no results or error
   */
  public List<String> executeQuery(String sqlQuery) {
    log.debug("Executing SQL query: {}", sqlQuery);

    try {
      String containerName = detectPostgresContainer();
      ProcessBuilder processBuilder =
          new ProcessBuilder(
              "docker",
              "exec",
              containerName,
              "psql",
              "-U",
              USER,
              "-d",
              DATABASE,
              "-t",
              "-A",
              "-c",
              sqlQuery);
      processBuilder.redirectErrorStream(true);

      Process process = processBuilder.start();

      List<String> results = new ArrayList<>();
      StringBuilder output = new StringBuilder();
      try (BufferedReader reader =
          new BufferedReader(new InputStreamReader(process.getInputStream()))) {
        String line;
        while ((line = reader.readLine()) != null) {
          output.append(line).append("\n");
          String trimmed = line.trim();
          if (!trimmed.isEmpty()) {
            results.add(trimmed);
          }
        }
      }

      int exitCode = process.waitFor();
      if (exitCode != 0) {
        log.warn(
            "SQL query failed with exit code: {} (container: {}). Output:\n{}",
            exitCode,
            containerName,
            output);
        return new ArrayList<>();
      }

      log.debug("Query returned {} rows", results.size());
      return results;
    } catch (Exception e) {
      log.warn("Failed to execute SQL query: {}", e.getMessage());
      return new ArrayList<>();
    }
  }

  /**
   * Executes a SQL query and returns the first result value as a string.
   *
   * @param sqlQuery SQL query to execute (should return single value)
   * @return First result value, or null if no results or error
   */
  public String executeQuerySingleValue(String sqlQuery) {
    List<String> results = executeQuery(sqlQuery);
    return results.isEmpty() ? null : results.get(0);
  }

  /**
   * Executes a SQL update/delete/insert statement.
   *
   * @param sqlStatement SQL statement to execute
   * @return true if successful, false otherwise
   */
  public boolean executeUpdate(String sqlStatement) {
    log.debug("Executing SQL update: {}", sqlStatement);

    try {
      String containerName = detectPostgresContainer();
      ProcessBuilder processBuilder =
          new ProcessBuilder(
              "docker",
              "exec",
              containerName,
              "psql",
              "-U",
              USER,
              "-d",
              DATABASE,
              "-c",
              sqlStatement);
      processBuilder.redirectErrorStream(true);

      Process process = processBuilder.start();

      // Read output for debugging
      StringBuilder output = new StringBuilder();
      try (BufferedReader reader =
          new BufferedReader(new InputStreamReader(process.getInputStream()))) {
        String line;
        while ((line = reader.readLine()) != null) {
          output.append(line).append("\n");
        }
      }

      int exitCode = process.waitFor();
      if (exitCode != 0) {
        log.error(
            "SQL update failed with exit code: {} (container: {}): {}",
            exitCode,
            containerName,
            output);
        return false;
      }

      log.debug("SQL update successful");
      return true;
    } catch (Exception e) {
      log.error("Failed to execute SQL update: {}", e.getMessage(), e);
      return false;
    }
  }

  /**
   * Checks enrollment status in database.
   *
   * @param enrollmentId Enrollment ID to check
   * @return Enrollment status (CREATED, BOUND, VERIFIED, INVALID) or null if not found
   */
  public String getEnrollmentStatus(Integer enrollmentId) {
    String sqlQuery =
        String.format(
            "SELECT enrollment_status FROM ezkey_enrollment WHERE enrollment_id = %d;",
            enrollmentId);
    return executeQuerySingleValue(sqlQuery);
  }

  /**
   * Gets enrollment proof token hash from database.
   *
   * @param enrollmentId Enrollment ID to check
   * @return Enrollment proof token hash (SHA-256 hex) or null if not found
   */
  public String getEnrollmentProofTokenHash(Integer enrollmentId) {
    String sqlQuery =
        String.format(
            "SELECT enrollment_proof_token_hash FROM ezkey_enrollment WHERE enrollment_id = %d;",
            enrollmentId);
    return executeQuerySingleValue(sqlQuery);
  }

  /**
   * Finds integration ID by name (i18n).
   *
   * <p>Note: This is a simple lookup. For more complex queries, use executeQuery directly.
   *
   * @param integrationName Integration name to find
   * @return Integration ID, or null if not found
   */
  public Integer findIntegrationIdByName(String integrationName) {
    String sqlQuery =
        String.format(
            "SELECT i.integration_id FROM ezkey_integration i "
                + "JOIN ezkey_integration_i18n i18n ON i.integration_id = i18n.integration_id "
                + "WHERE i18n.integration_i18n_name = '%s' LIMIT 1;",
            integrationName.replace("'", "''")); // SQL injection protection
    String result = executeQuerySingleValue(sqlQuery);
    return result != null ? Integer.parseInt(result.trim()) : null;
  }

  /**
   * Resets enrollment to CREATED state (clears device binding).
   *
   * <p>This is appropriate for functional E2E tests in clean room Docker environment.
   *
   * @param enrollmentId Enrollment ID to reset
   * @throws IllegalStateException if reset fails
   */
  public void resetEnrollment(Integer enrollmentId) {
    log.info("Resetting enrollment {} in database to CREATED state...", enrollmentId);

    String sqlUpdate =
        String.format(
            "UPDATE ezkey_enrollment SET enrollment_status = 'CREATED', device_public_key = NULL,"
                + " device_public_key_hash = NULL WHERE enrollment_id = %d;",
            enrollmentId);

    boolean success = executeUpdate(sqlUpdate);
    if (!success) {
      log.error("Failed to reset enrollment {} in database", enrollmentId);
      throw new IllegalStateException(
          "Failed to reset enrollment " + enrollmentId + " in database");
    }
    log.info("✅ Enrollment {} reset successfully in database", enrollmentId);
  }

  /**
   * Deletes integration and all related data (cascade).
   *
   * <p><b>Warning:</b> This deletes all related enrollments, API keys, etc. Use with caution.
   *
   * @param integrationId Integration ID to delete
   * @return true if successful, false otherwise
   */
  public boolean deleteIntegration(Integer integrationId) {
    String sqlDelete =
        String.format("DELETE FROM ezkey_integration WHERE integration_id = %d;", integrationId);
    return executeUpdate(sqlDelete);
  }
}

File: ezkey-cli-python/tests/util/database_helper.py
"""
Ezkey - Open Source MFA/Passkey Alternative

Copyright (c) 2025 Ezkey contributors
Licensed under the MIT License. See LICENSE file in the project root for full license information.

CLI Component: Database Helper
Description: Helper for opportunistic database access in CLI integration tests

This is a Python port of the Java DatabaseHelper used in functional tests.
Provides methods to query and update PostgreSQL database directly via docker exec,
avoiding API authentication requirements when appropriate.

Usage Context: Use for:
- State verification (faster than API calls)
- Finding existing entities before creating new ones
- Cleanup operations
- Resetting state for test idempotence

Database Configuration: Assumes PostgreSQL container named "ezkey-postgres"
with database "ezkey_db" and user "postgres".

@since 2025
"""

import logging
import subprocess
from typing import List, Optional

logger = logging.getLogger(__name__)


class DatabaseHelper:
    """Helper class for opportunistic database access in CLI integration tests."""

    DOCKER_CONTAINER = "ezkey-postgres"
    DATABASE = "ezkey_db"
    USER = "postgres"

    def execute_query(self, sql_query: str) -> List[str]:
        """
        Execute SQL query and return results as list of strings (one per row).

        Args:
            sql_query: SQL query to execute

        Returns:
            List of result rows (trimmed strings), empty list if no results or error
        """
        logger.debug("Executing SQL query: %s", sql_query)

        try:
            process = subprocess.run(
                [
                    "docker",
                    "exec",
                    self.DOCKER_CONTAINER,
                    "psql",
                    "-U",
                    self.USER,
                    "-d",
                    self.DATABASE,
                    "-t",
                    "-A",
                    "-c",
                    sql_query,
                ],
                capture_output=True,
                text=True,
                check=False,
            )

            if process.returncode != 0:
                logger.warning("SQL query failed with exit code: %d", process.returncode)
                logger.debug("Error output: %s", process.stderr)
                return []

            # Parse results (one per line, trimmed)
            results = [line.strip() for line in process.stdout.splitlines() if line.strip()]
            logger.debug("Query returned %d rows", len(results))
            return results

        except Exception as e:
            logger.warning("Failed to execute SQL query: %s", e)
            return []

    def execute_query_single_value(self, sql_query: str) -> Optional[str]:
        """
        Execute SQL query and return first result value as string.

        Args:
            sql_query: SQL query to execute (should return single value)

        Returns:
            First result value, or None if no results or error
        """
        results = self.execute_query(sql_query)
        return results[0] if results else None

    def execute_update(self, sql_statement: str) -> bool:
        """
        Execute SQL update/delete/insert statement.

        Args:
            sql_statement: SQL statement to execute

        Returns:
            True if successful, False otherwise
        """
        logger.debug("Executing SQL update: %s", sql_statement)

        try:
            process = subprocess.run(
                [
                    "docker",
                    "exec",
                    self.DOCKER_CONTAINER,
                    "psql",
                    "-U",
                    self.USER,
                    "-d",
                    self.DATABASE,
                    "-c",
                    sql_statement,
                ],
                capture_output=True,
                text=True,
                check=False,
            )

            if process.returncode != 0:
                logger.error("SQL update failed with exit code: %d: %s", process.returncode, process.stderr)
                return False

            logger.debug("SQL update successful")
            return True

        except Exception as e:
            logger.error("Failed to execute SQL update: %s", e)
            return False

    def get_enrollment_status(self, enrollment_id: int) -> Optional[str]:
        """
        Get enrollment status from database.

        Args:
            enrollment_id: Enrollment ID to check

        Returns:
            Enrollment status (CREATED, BOUND, VERIFIED, INVALID) or None if not found
        """
        sql_query = f"SELECT enrollment_status FROM ezkey_enrollment WHERE enrollment_id = {enrollment_id};"
        return self.execute_query_single_value(sql_query)

    def get_enrollment_proof_token_hash(self, enrollment_id: int) -> Optional[str]:
        """
        Get enrollment proof token hash from database.

        Args:
            enrollment_id: Enrollment ID to check

        Returns:
            Enrollment proof token hash (SHA-256 hex) or None if not found
        """
        sql_query = (
            f"SELECT enrollment_proof_token_hash FROM ezkey_enrollment "
            f"WHERE enrollment_id = {enrollment_id};"
        )
        return self.execute_query_single_value(sql_query)

    def find_integration_id_by_name(self, integration_name: str) -> Optional[int]:
        """
        Find integration ID by name (i18n).

        Note: This is a simple lookup. For more complex queries, use execute_query directly.

        Args:
            integration_name: Integration name to find

        Returns:
            Integration ID, or None if not found
        """
        # SQL injection protection
        safe_name = integration_name.replace("'", "''")
        sql_query = (
            f"SELECT i.integration_id FROM ezkey_integration i "
            f"JOIN ezkey_integration_i18n i18n ON i.integration_id = i18n.integration_id "
            f"WHERE i18n.integration_i18n_name = '{safe_name}' LIMIT 1;"
        )
        result = self.execute_query_single_value(sql_query)
        return int(result) if result else None

    def delete_integration(self, integration_id: int) -> bool:
        """
        Delete integration and all related data (cascade).

        Warning: This deletes all related enrollments, API keys, etc. Use with caution.

        Args:
            integration_id: Integration ID to delete

        Returns:
            True if successful, False otherwise
        """
        sql_delete = f"DELETE FROM ezkey_integration WHERE integration_id = {integration_id};"
        return self.execute_update(sql_delete)

    def get_integration_active_status(self, integration_id: int) -> Optional[str]:
        """
        Get integration active status from database.

        Args:
            integration_id: Integration ID to check

        Returns:
            Active status ('true' or 'false') or None if not found
        """
        sql_query = f"SELECT active FROM ezkey_integration WHERE integration_id = {integration_id};"
        return self.execute_query_single_value(sql_query)

